File: w16/yjkwon/단어_변환.py
# 단어 변환

"""
백트래킹을 이용합니다.
현재까지 사용하지 않은 단어 중 철자가 하나 다른 단어를 사용해 다음 스텝을 진행합니다.
현재 단어가 타겟과 같다면 정답을 갱신합니다.
정답이 발견되지 않으면 0, 발견되면 정답을 리턴합니다.
"""


def solution(begin, target, words):
    answer = len(words) + 1

    def back(visited, begin):
        nonlocal answer
        if begin == target:
            answer = min(len(visited) - 1, answer)

        for w in words:
            if w in visited:
                continue

            n = 0
            for a, b in zip(begin, w):
                if a == b:
                    n += 1
            if n == len(begin) - 1:
                visited.add(w)
                back(visited, w)
                visited.remove(w)

    back(set([begin]), begin)

    return answer if answer < len(words) + 1 else 0

File: w16/yongseon/단어변환.java
package w16.yongseon;

import java.util.*;

public class 단어변환 {
    public class Node {
        String word;
        Integer step;

        public Node(String word, Integer step) {
            this.word = word;
            this.step = step;
        }
    }

    private static boolean[] visited;
    private static Queue<Node> q = new LinkedList<>();

    public int solution(String begin, String target, String[] words) {
        visited = new boolean[words.length];
        q.add(new Node(begin, 0));

        while (!q.isEmpty()) {
            Node currentNode = q.poll();

            if (currentNode.word.equals(target)) {
                return currentNode.step;
            }

            for (int i = 0; i < words.length; i++) {
                int count = 0;

                if (!visited[i] && isTransform(currentNode.word, words[i])) {
                    visited[i] = true;
                    q.add(new Node(words[i], currentNode.step + 1));
                }
            }
        }

        return 0;
    }

    private boolean isTransform(String word, String targetWord) {
        int diffWord = 0;

        for (int i = 0; i < word.length(); i++) {
            if (word.charAt(i) != targetWord.charAt(i)) {
                diffWord++;

                if (diffWord > 1) return false;
            }
        }

        return true;
    }

}
